### ISSUE TITLE
How to write a response in chunks (USB control read)?

### ISSUE THREAD

**Issue by DurandA** [whitequark/libfx2#6]
I am trying to write large responses to a USB control read. The data is read in chunks of 32 bytes from a i2c device and feeds `EP0BUF`.

I had a look at something similar in the [Glasgow firmware](https://github.com/GlasgowEmbedded/glasgow/blob/85c55e56c650cad606370a55b9eaac6f5249c28b/firmware/main.c#L449-L471) and wrote this:
```C
uint8_t USB_REQ_ATECC_READ = 0x20;

if((req->bmRequestType == (USB_RECIP_DEVICE|USB_TYPE_VENDOR|USB_DIR_IN)) &&
   req->bRequest == USB_REQ_ATECC_READ) {
  uint8_t arg_slot = req->wValue;
  uint8_t arg_offset = req->wIndex;
  uint16_t arg_len = req->wLength;
  pending_setup = false;

  while(arg_len > 0) {
    uint8_t chunk_len = arg_len < 32 ? arg_len : 32;

    while(EP0CS & _BUSY);
    memset(EP0BUF, 0xFF, chunk_len); // whatever
    SETUP_EP0_BUF(chunk_len);

    arg_len  -= chunk_len;
    arg_offset += 1;
  }

  return;
}
```

I removed the i2c-related code so I just expect the FX2 to send garbage data. However, when I request more than 32 bytes (e.g. 33), I always receive 32 bytes:

```python
data = await device.control_read(usb1.REQUEST_TYPE_VENDOR, REQ_ATECC_READ, 0, 0, 33)
print(len(data)) # == 32
```

In other cases, an `usb1.USBErrorPipe: LIBUSB_ERROR_PIPE [-9]` exception is thrown.

Sorry for opening an issue here as this is probably not an issue with the _libfx2_ implementation.

**Comment by whitequark** [comment]
I suggest using Wireshark to look at the requests and then comparing this with the information provided in the FX2 manual.

**Comment by DurandA** [comment]
Apparently I can only send chunks of 64 bytes in EP0 if the data does not fit in a single packet. I tried with a few other values (32, 63) but this always results in an error.

> I suggest using Wireshark to look at the requests

That was also my first idea. Unfortunately, the packets are captured by libusb which inserts its own errors before propagating to Wireshark.

> and then comparing this with the information provided in the FX2 manual.

I think 64 bytes packets are expected with the `SDPAUTO=1`. I did not try it yet but I believe smaller packets are possible by changing `EP0BCL`.

Thanks @whitequark for your awesome projects. If you think that this behavior is expected, feel free to close the issue.
